SpatialGDK/Build/Programs/Improbable.Unreal.Scripts/WorkerCoordinator/LifetimeComponent.cs

```diff
@@ -0,0 +1,208 @@
+        {
+            numArgs = 0;
+
+            // Max lifetime argument.
+            int maxLifetime = 0;
+            if (Util.HasIntegerArgument(args, MaxLifetimeArg))
+            {
+                maxLifetime = Util.GetIntegerArgument(args, MaxLifetimeArg);
+                numArgs++;
+            }
+
+            // Use max lifetime as default.
+            int minLifetime = maxLifetime;
+            if (Util.HasIntegerArgument(args, MinLifetimeArg))
+            {
+                minLifetime = Util.GetIntegerArgument(args, MinLifetimeArg);
+                numArgs++;
+            }
+
+            // Use 60 seconds as default.
+            int restartAfterSeconds = 60;
+            if (Util.HasIntegerArgument(args, RestartAfterSecondsArg))
+            {
+                restartAfterSeconds = Util.GetIntegerArgument(args, RestartAfterSecondsArg);
+                numArgs++;
+            }
+
+            // Default do not use new simulated player to restart.
+            int useNewSimulatedPlayer = 0;
+            if (Util.HasIntegerArgument(args, UseNewSimulatedPlayerArg))
+            {
+                useNewSimulatedPlayer = Util.GetIntegerArgument(args, UseNewSimulatedPlayerArg);
+                numArgs++;
+            }
+
+            // Disable function by do not define max lifetime.
+            if (maxLifetime > 0)
+            {
+                return new LifetimeComponent(maxLifetime, minLifetime, restartAfterSeconds, useNewSimulatedPlayer > 0, logger);
+            }
+
+            return null;
+        }
+
+        private LifetimeComponent(int maxLifetime, int minLifetime, int restartAfterSeconds, bool useNewSimulatedPlayer, Logger logger)
+        {
+            UseNewSimulatedPlayer = useNewSimulatedPlayer;
+            MaxLifetime = maxLifetime;
+            MinLifetime = minLifetime;
+            RestartAfterSeconds = restartAfterSeconds;
+            Logger = logger;
+
+            WaitingList = new List<ClientInfo>();
+            RunningList = new List<ClientInfo>();
+
+            Random = new Random(Guid.NewGuid().GetHashCode());
+        }
+
+        public void SetHost(ILifetimeComponentHost host)
+        {
+            Host = host;
+        }
+
+        public void AddSimulatedPlayer(ClientInfo clientInfo)
+        {
+            WaitingList.Add(clientInfo);
+
+            Logger.WriteLog($"[LifetimeComponent][{DateTime.Now.ToString("HH:mm:ss")}] Add client ClientName={clientInfo.ClientName}.");
+        }
+
+        private long NewLifetimeTicks()
+        {
+            return TimeSpan.FromMinutes(Random.Next(MinLifetime, MaxLifetime)).Ticks;
+        }
+
+        public void Start()
+        {
+            // Loop tick.
+            while (true)
+            {
+                Tick();
+
+                Thread.Sleep(TimeSpan.FromSeconds(TickIntervalSeconds));
+            }
+        }
+
+        private void Tick()
+        {
+            CurTicks = DateTime.Now.Ticks;
```

Comment:
Is there a reason we use Ticks here instead of DateTime?

Reply:
What we want is `Ticks`, we do not need a struct here. 
It is easier to compare like `if (CurTicks >= SimulatedClientInfo.StartTick)`.
Not `if (CurDateTime.Ticks >= SimulatedClientInfo.StartTick)`.

Reply:
These would need to be updated, as well - we'll have 
```
DateTime CurTime = DateTime.Now;

// ...

if (CurTime >= SimulatedClientInfo.EndTime)
{
	// ...
}
```
The struct would be updated to
```
struct ClientInfo
    {
        public string ClientName;
        public string DevAuthToken;
        public string TargetDeployment;
        public DateTime StartTick;
        public DateTime EndTick;
    }
```
We won't need to use Ticks at all in this case.

Reply:
Yes, `DateTime` works here too. I don't choose `DateTime` because I just want to compare time, all other data inside `DateTime` is useless for me. Unless saving `DateTime` gives more benefits.

Reply:
The only benefit I can think of is clarity; essentially, you're still operating on `DateTime` and `TimeSpan` here, always converting them to `Ticks`, which may obscure the purpose somewhat. As in, here https://github.com/spatialos/UnrealGDK/pull/2935/files#diff-b6b504835d0a840e95597d99c44537d4031cfc3353d8ee29322f435732efdc68R224 you want to set `EndTicks` to the time when a client should be closed; doing `SimulatedClientInfo.EndTime = CurrentTime + NewLifetime();` would be the same, without converting everything to Ticks.

Reply:
Change it to `clientInfo.EndTime = curTime.AddMinutes(Random.Next(MinLifetime, MaxLifetime));`